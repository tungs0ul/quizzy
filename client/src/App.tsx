import { useState, useEffect } from "react";
import useWebSocket from "react-use-websocket";
import useLocalStorage from "./hooks/useLocalStorage";
import StudentView from "./components/views/StudentView";
import AdminView from "./components/views/AdminView";
import { BrowserRouter, Routes, Route, Link } from "react-router-dom";

// const WSURI = "ws://localhost:8080/ws";
const WSURI =
  ((window.location.protocol == "https:" && "wss://") || "ws://") +
  window.location.host +
  "/ws";

const initQuestion = {
  msg: "question",
  data: "",
  options: [],
};

export type Question = {
  type: string;
  data: string;
  options: string[];
};

export type Result = {
  name: string;
  value: number;
};

function App() {
  const { sendMessage, lastMessage } = useWebSocket(WSURI, {
    shouldReconnect: () => true,
    reconnectInterval: 5000,
  });
  const [question, setQuestion] = useLocalStorage("question", initQuestion);
  const [answer, setAnswer] = useLocalStorage("answer", null);
  const [result, setResult] = useState<Result[]>([]);
  const [notice, setNotice] = useState("");
  const [active, setActive] = useState(false);
  const [intervalID, setIntervalID] = useState<number | null>(null);

  useEffect(() => {
    if (lastMessage !== null) {
      if (!lastMessage.data) {
        return;
      }
      const data = JSON.parse(lastMessage.data);
      if (data.type === "question") {
        if (data.data !== question.data) {
          setNotice("The question has been changed!");
          setAnswer(null);
          setQuestion(data);
          setResult([]);
          if (intervalID !== null) {
            clearInterval(intervalID);
          }
          let id = setInterval(() => setNotice(""), 1500);
          setIntervalID(id);
        }
      } else if (data.type === "result") {
        if (
          data.data.reduce((acc: number, cur: number) => acc + cur, 0) === 0
        ) {
          setAnswer(null);
        }
        let result: Result[] = [];
        data.data.forEach((e: number, i: number) => {
          result.push({ name: question.options[i], value: e });
        });
        setResult(result);
      } else if (data.type === "clear") {
        setAnswer(null);
        setNotice("The answer has been cleared!");
        setInterval(() => setNotice(""), 1500);
      } else if (data.type === "reset") {
        setQuestion(initQuestion);
        setAnswer(null);
        setResult([]);
      }
    }
  }, [lastMessage]);

  return (
    <BrowserRouter>
      <div className="h-screen relative">
        <div className="flex justify-between items-center pb-4 pr-10 header">
          <div className="cursor-pointer p-4 text-yellow-600 hover:text-green-600 text-4xl items-center h-full flex justify-center">
            <Link to="/login">Admin</Link>
          </div>
          <div className="flex justify-end tablet:text-12xl laptop:text-16xl mobile:text-6xl font-bold">
            QUIZZY
          </div>
        </div>
        <Routes>
          <Route
            path="/"
            element={
              <StudentView
                answer={answer}
                question={question}
                result={result}
                sendMessage={sendMessage}
                setAnswer={setAnswer}
                notice={notice}
              />
            }
          />
          <Route
            path="/login"
            element={
              <AdminView
                sendMessage={sendMessage}
                setResult={setResult}
                result={result}
                question={question}
              />
            }
          />
        </Routes>
      </div>
    </BrowserRouter>
  );
}

export default App;
