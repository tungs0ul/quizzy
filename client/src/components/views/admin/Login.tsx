import React, { useState } from "react";
import Axios from "axios";

type Props = {
  setIsAdmin: any;
};

const API = window.location.protocol + "//" + window.location.host + "/admin";
// const API = "http://localhost:8080/admin";

export default function Login({ setIsAdmin }: Props) {
  const [password, setPassword] = useState("");
  const handleLogin = () => {
    Axios.post(API, { password }).then((res) => {
      if (res.data === "Success") {
        setIsAdmin(true);
      } else {
        alert("Wrong password!");
        setPassword("");
      }
    });
  };

  return (
    <div className="p-8 flex flex-col gap-4">
      <div className="flex justify-center text-3xl">
        <input
          type="password"
          className="h-16 p-2 text-right"
          value={password}
          onChange={(e) => setPassword(e.target.value)}
        />
      </div>
      <div className="flex justify-center text-4xl font-bold">
        <div
          className="border-2 border-blue-500 p-4 rounded-md hover:text-green-800 cursor-pointer hover:border-yellow-400"
          onClick={handleLogin}
        >
          Login
        </div>
      </div>
    </div>
  );
}
