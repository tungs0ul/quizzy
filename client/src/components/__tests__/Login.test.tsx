import TestRenderer, { act } from "react-test-renderer";
import { Result } from "../../App";
import * as React from "react";
import Login from "../views/admin/Login";

jest.mock("axios");
const setIsAdmin = jest.fn();

it("render Login", () => {
  const test = TestRenderer.create(<Login setIsAdmin={setIsAdmin} />);
  const testInstance = test.root;
  expect(test.toJSON()).toMatchSnapshot();

  const mEvent = { target: { value: "teresa teng" } };
  act(() => {
    testInstance.findByType("input").props.onChange(mEvent);
  });
  expect(testInstance.findByType("input").props.value).toEqual("teresa teng");
});
