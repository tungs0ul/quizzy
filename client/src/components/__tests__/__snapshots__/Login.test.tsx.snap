// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Login 1`] = `
<div
  className="p-8 flex flex-col gap-4"
>
  <div
    className="flex justify-center text-3xl"
  >
    <input
      className="h-16 p-2 text-right"
      onChange={[Function]}
      type="password"
      value=""
    />
  </div>
  <div
    className="flex justify-center text-4xl font-bold"
  >
    <div
      className="border-2 border-blue-500 p-4 rounded-md hover:text-green-800 cursor-pointer hover:border-yellow-400"
      onClick={[Function]}
    >
      Login
    </div>
  </div>
</div>
`;
